File: src/main/java/ar/edu/utn/frba/tacs/tp/api/herocardsgame/models/game/Player.java
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game;

import java.util.ArrayList;
import java.util.List;

public class Player {

    private String username;
    private List<Card> availableCards;
    private List<Card> prizeCards;

    public Player(String username, List<Card> availableCards) {
        this.username = username;
        this.availableCards = availableCards;
        this.prizeCards = new ArrayList(); // Starts empty
    }

    public String getUsername() {
        return username;
    }

    public List<Card> getAvailableCards() {
        return availableCards;
    }

    public List<Card> getPrizeCards() {
        return prizeCards;
    }
}

File: src/main/kotlin/ar/edu/utn/frba/tacs/tp/api/herocardsgame/models/game/Player.kt
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game

import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.accounts.User

data class Player(
    val id: Long? = null,
    val user: User,
    val availableCards: List<Card> = emptyList(),
    val prizeCards: List<Card> = emptyList()
) {

    fun winDuel(priceCard: Card): Player =
        this.copy(
            prizeCards = prizeCards + priceCard + availableCards.first(),
            availableCards = availableCards.drop(1)
        )

    fun loseDuel(): Player =
        this.copy(availableCards = availableCards.drop(1))

    fun tieDuel(): Player =
        this.copy(
            prizeCards = prizeCards + availableCards.first(),
            availableCards = availableCards.drop(1)
        )

    fun winMatch(): Player = copy(user = user.winMatch())

    fun tieMatch(): Player = copy(user = user.tieMatch())

    fun loseMatch(): Player = copy(user = user.loseMatch())

    fun startMatch(): Player = copy(user = user.startMatch())
}